File: python/comparison_engine.py
from typing import Dict, List, Any
from datetime import datetime

class ComparisonEngine:
    def __init__(self):
        self.insights = []
    
    def compare_runs(self, current_groups: Dict[str, List[Dict]], 
                    previous_groups: Dict[str, List[Dict]]) -> Dict[str, Any]:
        """Compare two grouped API runs and generate insights"""
        self.insights = []
        
        # Overall comparison
        current_total = sum(len(apis) for apis in current_groups.values())
        previous_total = sum(len(apis) for apis in previous_groups.values())
        
        # Failure analysis
        current_failures = self._count_failures(current_groups)
        previous_failures = self._count_failures(previous_groups)
        
        # Empty response analysis
        current_empty = self._count_empty_responses(current_groups)
        previous_empty = self._count_empty_responses(previous_groups)
        
        # Generate insights
        self._analyze_failures(current_groups, previous_groups)
        self._analyze_empty_responses(current_empty, previous_empty, current_total, previous_total)
        self._analyze_latency_by_group(current_groups, previous_groups)
        
        return {
            "timestamp": datetime.now().isoformat(),
            "summary": {
                "current_total": current_total,
                "previous_total": previous_total,
                "current_failures": current_failures,
                "previous_failures": previous_failures,
                "current_empty": current_empty,
                "previous_empty": previous_empty
            },
            "insights": self.insights,
            "group_analysis": self._analyze_groups(current_groups, previous_groups)
        }
    
    def _count_failures(self, groups: Dict[str, List[Dict]]) -> int:
        """Count total failures across all groups"""
        return sum(1 for apis in groups.values() 
                  for api in apis if api.get('status') == 'fail')
    
    def _count_empty_responses(self, groups: Dict[str, List[Dict]]) -> int:
        """Count total empty responses across all groups"""
        return sum(1 for apis in groups.values() 
                  for api in apis if api.get('empty_response', False) or api.get('isEmpty', False))
    
    def _analyze_failures(self, current_groups: Dict, previous_groups: Dict):
        """Analyze failure changes between runs"""
        current_failed_apis = set()
        previous_failed_apis = set()
        
        for group, apis in current_groups.items():
            for api in apis:
                if api.get('status') == 'fail':
                    url = api.get('api') or api.get('url', 'unknown')
                    current_failed_apis.add(url)
        
        for group, apis in previous_groups.items():
            for api in apis:
                if api.get('status') == 'fail':
                    url = api.get('api') or api.get('url', 'unknown')
                    previous_failed_apis.add(url)
        
        new_failures = current_failed_apis - previous_failed_apis
        recovered_apis = previous_failed_apis - current_failed_apis
        
        if new_failures:
            self.insights.append(f"{len(new_failures)} APIs failed today that passed yesterday")
        
        if recovered_apis:
            self.insights.append(f"{len(recovered_apis)} APIs recovered from previous failures")
    
    def _analyze_empty_responses(self, current_empty: int, previous_empty: int, 
                                current_total: int, previous_total: int):
        """Analyze empty response changes"""
        if previous_total == 0:
            return
        
        current_pct = (current_empty / current_total) * 100
        previous_pct = (previous_empty / previous_total) * 100
        change = current_pct - previous_pct
        
        if abs(change) > 5:
            direction = "up" if change > 0 else "down"
            self.insights.append(f"Empty responses {direction} {abs(change):.1f}% since last scan")
    
    def _analyze_latency_by_group(self, current_groups: Dict, previous_groups: Dict):
        """Analyze latency changes by endpoint group"""
        for group in current_groups:
            if group not in previous_groups:
                continue
            
            current_latencies = [api.get('latency_ms', api.get('latency', 0)) for api in current_groups[group]]
            previous_latencies = [api.get('latency_ms', api.get('latency', 0)) for api in previous_groups[group]]
            
            if not current_latencies or not previous_latencies:
                continue
            
            current_avg = sum(current_latencies) / len(current_latencies)
            previous_avg = sum(previous_latencies) / len(previous_latencies)
            
            if previous_avg == 0:
                continue
                
            change_pct = ((current_avg - previous_avg) / previous_avg) * 100
            
            if abs(change_pct) > 20:
                direction = "↑" if change_pct > 0 else "↓"
                self.insights.append(f"Latency {direction} {abs(change_pct):.0f}% in {group}")
    
    def _analyze_groups(self, current_groups: Dict, previous_groups: Dict) -> Dict:
        """Detailed group-by-group analysis"""
        analysis = {}
        
        all_groups = set(current_groups.keys()) | set(previous_groups.keys())
        
        for group in all_groups:
            current_apis = current_groups.get(group, [])
            previous_apis = previous_groups.get(group, [])
            
            analysis[group] = {
                "current_count": len(current_apis),
                "previous_count": len(previous_apis),
                "current_failures": sum(1 for api in current_apis if api.get('status') == 'fail'),
                "previous_failures": sum(1 for api in previous_apis if api.get('status') == 'fail'),
                "avg_latency_current": sum(api.get('latency_ms', api.get('latency', 0)) for api in current_apis) / len(current_apis) if current_apis else 0,
                "avg_latency_previous": sum(api.get('latency_ms', api.get('latency', 0)) for api in previous_apis) / len(previous_apis) if previous_apis else 0
            }
        
        return analysis

File: comparison-engine.js
class ComparisonEngine {
    constructor() {
        this.insights = [];
    }

    compare(currentSnapshot, previousSnapshot) {
        if (!previousSnapshot) {
            return {
                summary: "📊 First run - no comparison available",
                insights: ["🆕 Baseline established with " + currentSnapshot.summary.totalAPIs + " APIs"],
                details: {}
            };
        }

        this.insights = [];
        const details = {};

        // Compare API failures
        const failureComparison = this.compareFailures(currentSnapshot, previousSnapshot);
        details.failures = failureComparison;

        // Compare empty responses
        const emptyComparison = this.compareEmptyResponses(currentSnapshot, previousSnapshot);
        details.emptyResponses = emptyComparison;

        // Compare latency
        const latencyComparison = this.compareLatency(currentSnapshot, previousSnapshot);
        details.latency = latencyComparison;

        // Compare total APIs
        const totalComparison = this.compareTotalAPIs(currentSnapshot, previousSnapshot);
        details.totalAPIs = totalComparison;

        return {
            summary: this.generateSummary(),
            insights: this.insights,
            details: details
        };
    }

    compareFailures(current, previous) {
        const currentFailed = this.getFailedApis(current);
        const previousFailed = this.getFailedApis(previous);
        
        const newFailures = currentFailed.filter(api => !previousFailed.includes(api));
        const recoveredApis = previousFailed.filter(api => !currentFailed.includes(api));
        
        if (newFailures.length > 0) {
            this.insights.push(`🔺 ${newFailures.length} APIs failed today that didn't yesterday`);
        }
        
        if (recoveredApis.length > 0) {
            this.insights.push(`✅ ${recoveredApis.length} APIs recovered from previous failures`);
        }

        return {
            current: currentFailed.length,
            previous: previousFailed.length,
            newFailures: newFailures,
            recovered: recoveredApis,
            change: currentFailed.length - previousFailed.length
        };
    }

    compareEmptyResponses(current, previous) {
        const currentEmpty = current.summary.emptyResponses;
        const previousEmpty = previous.summary.emptyResponses;
        const currentTotal = current.summary.totalAPIs;
        const previousTotal = previous.summary.totalAPIs;
        
        const currentPercent = (currentEmpty / currentTotal) * 100;
        const previousPercent = (previousEmpty / previousTotal) * 100;
        const percentChange = currentPercent - previousPercent;
        
        if (Math.abs(percentChange) > 5) {
            const direction = percentChange > 0 ? 'up' : 'down';
            this.insights.push(`📉 Empty responses ${direction} ${Math.abs(percentChange).toFixed(1)}% since last scan`);
        }

        return {
            current: currentEmpty,
            previous: previousEmpty,
            currentPercent: currentPercent,
            previousPercent: previousPercent,
            change: percentChange
        };
    }

    compareLatency(current, previous) {
        const currentAvg = current.summary.avgLatency;
        const previousAvg = previous.summary.avgLatency;
        const latencyChange = currentAvg - previousAvg;
        
        if (Math.abs(latencyChange) > 100) {
            const direction = latencyChange > 0 ? 'increased' : 'decreased';
            this.insights.push(`⚠️ Avg latency ${direction} by ${Math.abs(latencyChange).toFixed(0)}ms`);
        }

        // Find APIs with significant latency changes
        const slowApis = this.findSlowApis(current, previous);
        if (slowApis.length > 0) {
            this.insights.push(`🐌 ${slowApis.length} endpoints showing increased latency`);
        }

        return {
            current: currentAvg,
            previous: previousAvg,
            change: latencyChange,
            slowApis: slowApis
        };
    }

    compareTotalAPIs(current, previous) {
        const currentTotal = current.summary.totalAPIs;
        const previousTotal = previous.summary.totalAPIs;
        const change = currentTotal - previousTotal;
        
        if (change !== 0) {
            const direction = change > 0 ? 'increased' : 'decreased';
            this.insights.push(`📊 Total APIs ${direction} by ${Math.abs(change)}`);
        }

        return {
            current: currentTotal,
            previous: previousTotal,
            change: change
        };
    }

    getFailedApis(snapshot) {
        return snapshot.apis
            .filter(api => api.status === 'fail')
            .map(api => api.id);
    }

    findSlowApis(current, previous) {
        const slowApis = [];
        const previousApiMap = new Map(previous.apis.map(api => [api.id, api]));
        
        current.apis.forEach(currentApi => {
            const previousApi = previousApiMap.get(currentApi.id);
            if (previousApi && currentApi.latency > previousApi.latency + 200) {
                slowApis.push({
                    id: currentApi.id,
                    url: currentApi.url,
                    currentLatency: currentApi.latency,
                    previousLatency: previousApi.latency,
                    increase: currentApi.latency - previousApi.latency
                });
            }
        });
        
        return slowApis;
    }

    generateSummary() {
        if (this.insights.length === 0) {
            return "✅ No significant changes detected";
        }
        
        return `📈 ${this.insights.length} changes detected`;
    }

    displayComparison(comparison) {
        console.log('\n' + '='.repeat(50));
        console.log('📊 API COMPARISON REPORT');
        console.log('='.repeat(50));
        console.log(`\n${comparison.summary}\n`);
        
        if (comparison.insights.length > 0) {
            comparison.insights.forEach(insight => console.log(insight));
        }
        
        console.log('\n' + '-'.repeat(30));
        console.log('📋 DETAILED METRICS:');
        console.log(`   Total APIs: ${comparison.details.totalAPIs?.current || 0} (${comparison.details.totalAPIs?.change >= 0 ? '+' : ''}${comparison.details.totalAPIs?.change || 0})`);
        console.log(`   Failures: ${comparison.details.failures?.current || 0} (${comparison.details.failures?.change >= 0 ? '+' : ''}${comparison.details.failures?.change || 0})`);
        console.log(`   Empty Responses: ${comparison.details.emptyResponses?.current || 0} (${comparison.details.emptyResponses?.change >= 0 ? '+' : ''}${comparison.details.emptyResponses?.change?.toFixed(1) || 0}%)`);
        console.log(`   Avg Latency: ${comparison.details.latency?.current?.toFixed(0) || 0}ms (${comparison.details.latency?.change >= 0 ? '+' : ''}${comparison.details.latency?.change?.toFixed(0) || 0}ms)`);
        console.log('='.repeat(50));
    }
}

module.exports = ComparisonEngine;